realm/realm-library/src/main/java/io/realm/RealmCache.java

```diff
@@ -105,25 +107,29 @@ private RealmCache(RealmConfiguration config) {
      * @param realmClass class of {@link Realm} or {@link DynamicRealm} to be created in or gotten from the cache.
      * @return the {@link Realm} or {@link DynamicRealm} instance.
      */
-    static synchronized <E extends BaseRealm> E createRealmOrGetFromCache(RealmConfiguration configuration,
+    static <E extends BaseRealm> E createRealmOrGetFromCache(RealmConfiguration configuration,
             Class<E> realmClass) {
-        boolean isCacheInMap = true;
-        RealmCache cache = cachesMap.get(configuration.getPath());
-        if (cache == null) {
-            // Creates a new cache.
-            cache = new RealmCache(configuration);
-            // The new cache should be added to the map later.
-            isCacheInMap = false;
-
-            copyAssetFileIfNeeded(configuration);
-        } else {
-            // Throws the exception if validation failed.
-            cache.validateConfiguration(configuration);
+        RealmCache cache;
+        synchronized (RealmCache.class) {
+            cache = cachesMap.get(configuration.getPath());
+            if (cache == null) {
+                // Creates a new cache.
+                cache = new RealmCache();
+                cachesMap.put(configuration.getPath(), cache);
+            }
         }
 
-        RefAndCount refAndCount = cache.refAndCountMap.get(RealmCacheType.valueOf(realmClass));
+        return cache.doCreateRealmOrGetFromCache(configuration, realmClass);
+    }
+
+    private synchronized <E extends BaseRealm> E doCreateRealmOrGetFromCache(RealmConfiguration configuration,
```

Comment:
You don't seem to gain much by moving the synchronized method from `createRealmOrGetFromCache` to `doCreateRealmOrGetFromCache` ?